### PR TITLE
add default scaffolder metrics

### PR DESCRIPTION
**Pull request by spencerrichardhenry**
## Hey, I just made a Pull Request!

This is for adding a set of default scaffolder metrics to Backstage.

github issue:
https://github.com/backstage/backstage/issues/12453
#### :heavy_check_mark: Checklist

- [:heavy_check_mark:] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [:heavy_check_mark:] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/silver-games-yell.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-scaffolder-backend': patch
+---
+
+Added a set of default prometheus metrics around scaffolding
```
Comment (awanlin) [inline .changeset/silver-games-yell.md:5]: Could we include a list of what metrics will be available? Also, some details about the labels would be nice.

Might be worth having a link to the metrics guide in the contrib folder as I think some people won't have that setup.

For the Vale error I'm wondering if you need to capitalize "prometheus" this to make Vale happy or to add it to the list of words it knows about.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 2]
```diff
@@ -37,7 +37,7 @@ import {
   TaskSpecV1beta3,
   TaskStep,
 } from '@backstage/plugin-scaffolder-common';
-import { UserEntity } from '@backstage/catalog-model';
+import { createCounterMetric, createHistogramMetric } from '../../../../catalog-backend/src/util/metrics';
```
Comment (awanlin) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:40]: I'm wondering if these need to be properly exported in the `catalog-backend` and then you can add that package and import them from there here

**Review by awanlin — COMMENTED** [review]
Thanks for this contribution @spencerrichardhenry, I know many in the community will be really happy with this. 🚀 

I've left a few minor comments

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 2 above)
Comment (spencerrichardhenry) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:40]: Definitely, thx for the catch

Comment (spencerrichardhenry) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:40]: Do you think we should move that whole class to backend-common instead?

File: .changeset/silver-games-yell.md (hunk 1 above)
Comment (spencerrichardhenry) [reply .changeset/silver-games-yell.md]: Addressed, please review. Also I'm not in love with the metric names if you think there's something clearer. mostly I thought of 'task' as a 'Template' and 'step' as a 'Task' but bootstrapping isnt where I spend most of my time and i just tried to follow the class's convention

Comment (awanlin) [reply .changeset/silver-games-yell.md]: There's something odd going on with the formatting, somethings have bullets that shouldn't so it's hard to read. Here's what I'm seeing:

image

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 2 above)
Comment (awanlin) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Yeah, this gets into some territory that I'm not sure what the maintainers will want to do, they are on an offsite so we won't know till next week. If I was in your shoes I would just do the export for now as that shouldn't be too much work to undo if they don't want to go that route

File: .changeset/silver-games-yell.md (hunk 1 above)
Comment (awanlin) [reply .changeset/silver-games-yell.md]: Looks good now 👍

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 2 above)
Comment (awanlin) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts]: Same with this, looks good 👍

**Comment by awanlin** [comment]
OK, one more thing, sorry, you'll want to run: `yarn build:api-reports`. That will generate updated API Reports based on your changes where needed

**Comment by spencerrichardhenry** [comment]
>

File: .changeset/silver-games-yell.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
+---
+'@backstage/plugin-scaffolder-backend': patch
+'@backstage/plugin-catalog-backend': patch
+---
+
+Added a set of default Prometheus metrics around scaffolding. See below for a list of metrics and an explanation of their labels:
```
Comment (freben) [inline .changeset/silver-games-yell.md:6]: Need an extra newline here so that the metric appears on a line of its own

File: plugins/catalog-backend/api-report.md [hunk 4]
```diff
@@ -289,12 +297,32 @@ export const createCatalogPermissionRule: <TParams extends unknown[]>(
   rule: PermissionRule<Entity, EntitiesSearchFilter, 'catalog-entity', TParams>,
 ) => PermissionRule<Entity, EntitiesSearchFilter, 'catalog-entity', TParams>;
 
+// @alpha
+export function createCounterMetric<T extends string>(
+  config: CounterConfiguration<T>,
+): Counter<T>;
```
Comment (freben) [inline plugins/catalog-backend/api-report.md:303]: Hm, should we export these right now or should we wait to make custom metrics into a separate thing? @benjdlambert One drawback here is that we start exporting `prom-client` types which is a potential risk for future compat problem when we want to upgrade that library to a newer version.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 5]
```diff
@@ -53,10 +56,7 @@ type TemplateContext = {
     [stepName: string]: { output: { [outputName: string]: JsonValue } };
   };
   secrets?: Record<string, string>;
-  user?: {
-    entity?: UserEntity;
-    ref?: string;
-  };
+  user?: {};
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:59]: why is this change here?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 6]
```diff
@@ -199,11 +200,11 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
       additionalTemplateFilters: this.options.additionalTemplateFilters,
     });
 
+    let taskTrack;
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:203]: this declaration doesn't have to be outside of the try/catch does it?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 7]
```diff
@@ -362,3 +352,129 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
...
+        name: step.name,
+      });
+      stepTimer({ result: 'failed' });
+    }
+
+    async function skipFalsy() {
+      await task.emitLog(
+        `Skipping step ${step.id} because it's if condition was false`,
+        { stepId: step.id, status: 'skipped' },
+      );
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:466]: Is prometheus ok with leaving started timers just hanging? Do we need to cancel `stepTimer` here or should we do `stepTimer({ result: 'skipped' });` (which of course is a bit useless too because it'll usually be like zero milliseconds every time)?

**Review by freben — COMMENTED** [review]
Alright thanks! Some feedback and some open questions.

File: plugins/catalog-backend/api-report.md (hunk 4 above)
Comment (benjdlambert) [reply plugins/catalog-backend/api-report.md:303]: I think we probably want to copy these for now to be honest. When the backend system work is getting shipped I'd expect that we have a `metricsApiRef` or something that everyone can use and we can provide a default implementation which uses `prom-client` under the hood.

Comment (benjdlambert) [reply plugins/catalog-backend/api-report.md:303]: I also dislike the `scaffolder-backend` depending on the `catalog-backend` at all, we should move the things like move the `CatalogProcessor` etc imports to `catalog-node` instead.

Comment (freben) [reply plugins/catalog-backend/api-report.md:303]: Oh wait, I'm blind - thanks for calling this out, I missed somehow that this was in catalog-backend and this is an absolute no-no and there is no reason to use any metrics export from the catalog. Building some facility into backend-common for declaring metrics isn't quite on the map right now, so the remaining option is indeed to copy this over into the scaffolder backend

Comment (awanlin) [reply plugins/catalog-backend/api-report.md:303]: Thanks @freben and @benjdlambert I wasn't really sure what direction to suggest @spencerrichardhenry should go with this

File: .changeset/silver-games-yell.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+- name
+
+scaffolder_step_duration: a histogram which tracks the duration of each step run
+
+- name
+
+- result
+
+label explanations:
```
[the remark this answers is not in the document]
Comment (spencerrichardhenry) [reply .changeset/silver-games-yell.md:39]: Would you like the definition with each instance or just the first?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 9]
```diff
@@ -362,3 +352,129 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
...
+  async function taskStart(task: TaskContext) {
+    await task.emitLog(`Starting up task with ${task.spec.steps.length} steps`);
+    const template = task.spec.templateInfo?.entity?.metadata.name || '';
+    const invoker = task.spec.user?.entity?.metadata.name || '';
+
+    const taskTimer = taskDuration.startTimer({
+      template,
+    });
+
+    async function skipDryRun(
```
[the remark this answers is not in the document]
Comment (spencerrichardhenry) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:397]: skipDryRun is called on line 238

File: .changeset/silver-games-yell.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (freben) [reply .changeset/silver-games-yell.md]: Each! Make it trivial to read

**Comment by freben** [comment]
I see that DCO is complaining. A hint - merging master into your own branch is usually way harder to get to working, rather than using rebase instead. Any breakage in the linear commit history is a risk for running into DCO troubles.

**Comment by freben** [comment]
I also see that CI complains about `prettier` errors, and indeed there's some oddness around line breaks etc when looking at the diff. You may want to enable auto format on save in your editor.

File: plugins/scaffolder-backend/package.json [hunk 10]
```diff
@@ -71,6 +71,7 @@
     "octokit": "^2.0.0",
     "octokit-plugin-create-pull-request": "^3.10.0",
     "p-limit": "^3.1.0",
+    "prom-client": "^14.1.0",
```
Comment (freben) [inline plugins/scaffolder-backend/package.json:74]: Sorry to be a pest, but could we make this `^14.0.1`? That's what the rest of the repo uses, and we don't want to get unintended duplicates.

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts (hunk 6 above)
Comment (freben) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:207]: That's in an inner catch block inside the try though, or maybe i'm misreading it

File: .changeset/silver-games-yell.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    Labels:
+
+    - `template`: The entity ref of the scaffolded template
+    - `result`: A boolean describing whether the task ran successfully
+
+ - scaffolder_step_success_count: a count that tracks each step run
+
+    Labels:
+
+    - `name`: The name of the step that was run
```
Comment (freben) [inline .changeset/silver-games-yell.md:33]: should update these three to have a `template` label too now. And could we make `name` be `step` instead?

File: plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts [hunk 12]
```diff
@@ -362,3 +356,132 @@ export class NunjucksWorkflowRunner implements WorkflowRunner {
...
+      name: step.name,
+    });
+
+    async function markSuccessful() {
+      await task.emitLog(`Finished step ${step.name}`, {
+        stepId: step.id,
+        status: 'completed',
+      });
+      stepSuccesses.inc({
+        name: step.name,
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:456]: forgot to pass in template to these three

Comment (spencerrichardhenry) [reply plugins/scaffolder-backend/src/scaffolder/tasks/NunjucksWorkflowRunner.ts:456]: timer had it, added to success and error

File: plugins/scaffolder-backend/package.json (hunk 10 above)
Comment (spencerrichardhenry) [reply plugins/scaffolder-backend/package.json:74]: not a pest at all, thanks for the catch

**Comment by spencerrichardhenry** [comment]
@freben I cannot figure out how to resolve the DCO error complaining about commit `bb943d7`.  rebasing through 130 commits is getting to be merging/history hell. Unless you know a neat solution here I might just make a clean PR with my changes squashed into one commit

**Comment by freben** [comment]
OK let's keep the other one then.
